vpr/src/route/connection_box_lookahead_map.cpp

```diff
@@ -537,6 +751,91 @@ float ConnectionBoxMapLookahead::get_expected_cost(
     }
 }
 
+static vtr::Rect<int> bounding_box_for_node(const t_rr_node& node) {
+    return vtr::Rect<int>(node.xlow(), node.ylow(),
+                          node.xhigh() + 1, node.yhigh() + 1);
+}
+
+static vtr::Point<int> point_for_node(const t_rr_node& node) {
+    int x = (node.xhigh() + node.xlow()) / 2;
+    int y = (node.yhigh() + node.ylow()) / 2;
+    return vtr::Point<int>(x, y);
+}
+
+// for each segment type, find the nearest nodes to an equally spaced grid of points
+// within the bounding box for that segment type
+static void find_inodes_for_segment_types(std::vector<SampleGrid>* inodes_for_segment) {
+    auto& device_ctx = g_vpr_ctx.device();
+    auto& rr_nodes = device_ctx.rr_nodes;
+    const int num_segments = inodes_for_segment->size();
+
+    // compute bounding boxes for each segment type
+    std::vector<vtr::Rect<int>> bounding_box_for_segment(num_segments, vtr::Rect<int>());
+    for (size_t i = 0; i < rr_nodes.size(); i++) {
+        auto& node = rr_nodes[i];
+        if (node.type() != CHANX && node.type() != CHANY) continue;
+        int seg_index = device_ctx.rr_indexed_data[node.cost_index()].seg_index;
+
+        VTR_ASSERT(seg_index != OPEN);
+        VTR_ASSERT(seg_index < num_segments);
+
+        bounding_box_for_segment[seg_index] |= bounding_box_for_node(node);
+    }
+
+    // select an inode near the center of the bounding box for each segment type
+    inodes_for_segment->clear();
+    inodes_for_segment->resize(num_segments);
+    for (auto& grid : *inodes_for_segment) {
+        for (int y = 0; y < SAMPLE_GRID_SIZE; y++) {
+            for (int x = 0; x < SAMPLE_GRID_SIZE; x++) {
+                grid.point[y][x].samples = std::vector<ssize_t>();
+            }
+        }
+    }
+
+    for (size_t i = 0; i < rr_nodes.size(); i++) {
+        auto& node = rr_nodes[i];
+        vtr::Rect<int> node_bounds = bounding_box_for_node(node);
+        if (node.type() != CHANX && node.type() != CHANY) continue;
+        if (node.capacity() == 0 || device_ctx.connection_boxes.find_canonical_loc(i) == nullptr) continue;
+
+        int seg_index = device_ctx.rr_indexed_data[node.cost_index()].seg_index;
+
+        VTR_ASSERT(seg_index != OPEN);
+        VTR_ASSERT(seg_index < num_segments);
+
+        auto& grid = (*inodes_for_segment)[seg_index];
+        for (int sy = 0; sy < SAMPLE_GRID_SIZE; sy++) {
```

Comment:
I think this loop needs refactoring, it is hard to understand because it combines steps.  The goto is totally unnecessary, and the code will be just as fast if you seperate the steps, and name them. 

In particular the two steps that should be split are:
- Identifying the sx/sy that contains the particular node.  This value is fixed (based on `sample`).  Deciding which sx/sy the node belongs too should be done first, and separate from whether the sample is used or not
- The logic for whether this is a "better" node than the previously collected node should be made into explicit logic (with comments), because the minimum distances to the center of the sample may not be the only metric we want to use.

Reply:
Is this still necessary?

Reply:
Absolutely?

Reply:
Done.